src/utils/gcsApi.ts

```diff
@@ -0,0 +1,321 @@
+            name: 'CRAFT_GCS_TARGET_CREDENTIALS_JSON',
+            legacyName: 'CRAFT_GCS_CREDENTIALS_JSON',
+          };
+    const filepathVar: RequiredConfigVar =
+      gcsRole === BucketRole.STORE
+        ? { name: 'CRAFT_GCS_STORE_CREDENTIALS_PATH' }
+        : {
+            name: 'CRAFT_GCS_TARGET_CREDENTIALS_PATH',
+            legacyName: 'CRAFT_GCS_CREDENTIALS_PATH',
+          };
+
+    // make sure we have at least one of the necessary variables
+    checkEnvForPrerequisite(jsonVar, filepathVar);
+
+    const gcsCredsJson = process.env[jsonVar.name];
+    const gcsCredsPath = process.env[filepathVar.name];
+
+    // necessary in case we're doing a dry run, in which case missing
+    // credentials will log an error but not throw
+    let configRaw = '';
+
+    if (gcsCredsJson) {
+      logger.debug(`Using configuration from ${jsonVar.name}`);
+      configRaw = gcsCredsJson;
+    } else if (gcsCredsPath) {
+      logger.debug(`Using configuration located at ${filepathVar.name}`);
+      if (!fs.existsSync(gcsCredsPath)) {
+        reportError(`File does not exist: ${gcsCredsPath}`);
+      }
+      configRaw = fs.readFileSync(gcsCredsPath).toString();
+    } else {
+      const errorMsg =
+        `GCS credentials not found! Please provide the path to the ` +
+        `configuration via environment variable ${filepathVar.name}, or ` +
+        `specify the entire configuration in ${jsonVar.name}.`;
+      reportError(errorMsg);
+    }
+
+    let creds: { [key: string]: string } = {};
+    try {
+      creds = JSON.parse(configRaw);
+    } catch (err) {
+      reportError('Error parsing JSON credentials');
+    }
+
+    for (const field of ['project_id', 'client_email', 'private_key']) {
+      if (!creds[field]) {
+        reportError(`GCS credentials missing ${field}!`);
+      }
+    }
+    return creds;
+  };
+
+  /**
+   * Detect the content-type based on regular expressions defined in
+   * CONTENT_TYPES_EXT.
+   *
+   * The underlying GCS package usually detects content-type itself, but it's
+   * not always correct.
+   *
+   * @param artifactName Name of the artifact to check
+   * @returns A content-type string, or undefined if the artifact name doesn't
+   * have a known extension
+   */
+  private detectContentType(artifactName: string): string | undefined {
+    for (const entry of CONTENT_TYPES_EXT) {
+      const [regex, contentType] = entry;
+      if (artifactName.match(regex)) {
+        return contentType;
+      }
+    }
+    return undefined;
+  }
+
+  /**
+   * Uploads the artifact at the given local path to the path specified in the
+   * given config object
+   *
+   * @param localFilePath Location of the file to be uploaded
+   * @param uploadConfig Configuration for the upload including destination path
+   * and metadata
+   */
+  private async uploadArtifactFromPath(
+    localFilePath: string,
+    uploadConfig: Pick<
+      Required<GCSUploadOptions>,
+      'destination' | 'metadata' | 'gzip'
```

Comment:
To clarify, you want to require those three attributes, correct?

Reply:
Yeah - I needed a way to translate the sentence

_You have to follow the `GCSUploadOptions` interface but also, even though I know these things are optional there, I need you to include `destination`, `metadata`, and `gip` in your object_.

into Typescript. If there's a more elegant way to do that, I'd definitely be interested to know.

Reply:
🤔 
Can't think of anything better.

Reply:
👍🏻